File: Other/R/results_viewer.R
# MSI Analysis Application - Results Viewer
# 結果可視化モジュール

#' 結果フォルダの構造を取得
#' @param result_dir 結果フォルダパス
#' @return フォルダ構造リスト
get_result_structure <- function(result_dir) {
  if (!dir.exists(result_dir)) {
    return(list())
  }
  
  structure <- list(
    root = result_dir,
    subdirs = list()
  )
  
  # 主要なサブディレクトリ
  key_dirs <- c("Harmony", "RPCA", "PCA", "Volcano_Plots", "Volcano_Plots_MRM", 
                "Cluster_Top5_MSI", "PerCluster_Highlight", "RDS_Files")
  
  for (subdir in key_dirs) {
    subdir_path <- file.path(result_dir, subdir)
    if (dir.exists(subdir_path)) {
      images <- list.files(subdir_path, pattern = "\\.(png|jpg|jpeg)$", 
                          full.names = TRUE, recursive = TRUE)
      structure$subdirs[[subdir]] <- list(
        path = subdir_path,
        image_count = length(images),
        images = images
      )
    }
  }
  
  # ルートディレクトリの画像
  root_images <- list.files(result_dir, pattern = "\\.(png|jpg|jpeg)$", 
                            full.names = TRUE, recursive = FALSE)
  structure$root_images <- root_images
  
  return(structure)
}

#' 画像カテゴリを判定
#' @param image_path 画像パス
#' @return カテゴリ名
categorize_image <- function(image_path) {
  filename <- tolower(basename(image_path))
  
  if (grepl("umap", filename)) return("UMAP")
  if (grepl("volcano", filename)) return("Volcano")
  if (grepl("msi|top5", filename)) return("MSI")
  if (grepl("spatial|cluster", filename)) return("Spatial")
  if (grepl("heatmap", filename)) return("Heatmap")
  if (grepl("tic", filename)) return("TIC")
  if (grepl("filter", filename)) return("Filtering")
  
  return("Other")
}

#' 画像一覧をカテゴリ別に整理
#' @param images 画像パスのベクトル
#' @return カテゴリ別リスト
organize_images_by_category <- function(images) {
  if (length(images) == 0) {
    return(list())
  }
  
  categories <- sapply(images, categorize_image)
  
  split(images, categories)
}

#' クラスタ番号を画像パスから抽出
#' @param image_path 画像パス
#' @return クラスタ番号（整数）またはNA
extract_cluster_number <- function(image_path) {
  filename <- basename(image_path)
  
  # パターン: Cluster_0, Cluster_10 など
  match <- regmatches(filename, regexpr("Cluster_\\d+", filename))
  if (length(match) > 0) {
    return(as.integer(gsub("Cluster_", "", match)))
  }
  
  # パターン: cluster_0, cluster_10 など
  match <- regmatches(filename, regexpr("cluster_\\d+", filename, ignore.case = TRUE))
  if (length(match) > 0) {
    return(as.integer(gsub("cluster_", "", match, ignore.case = TRUE)))
  }
  
  return(NA)
}

#' 結果フォルダ内の利用可能なクラスタ番号を取得
#' @param result_dir 結果フォルダパス
#' @return クラスタ番号のベクトル
get_available_clusters <- function(result_dir) {
  images <- list.files(result_dir, pattern = "\\.(png|jpg|jpeg)$", 
                       full.names = TRUE, recursive = TRUE)
  
  clusters <- unique(na.omit(sapply(images, extract_cluster_number)))
  sort(clusters)
}

#' サンプル名を画像パスから抽出
#' @param image_path 画像パス
#' @return サンプル名またはNA
extract_sample_name <- function(image_path) {
  filename <- basename(image_path)
  
  # 一般的なパターンを試行
  # 例: plot_cluster_harmony_251213_Kizu-Embryo-E16.png
  parts <- strsplit(filename, "_")[[1]]
  
  # 日付パターン（6桁の数字で始まる）を探す
  date_idx <- grep("^\\d{6}", parts)
  if (length(date_idx) > 0) {
    # 日付以降を結合してサンプル名とする
    sample_parts <- parts[date_idx[1]:length(parts)]
    sample_name <- paste(sample_parts, collapse = "_")
    sample_name <- gsub("\\.(png|jpg|jpeg)$", "", sample_name)
    return(sample_name)
  }
  
  return(NA)
}

#' 特定クラスタの画像をフィルタ
#' @param images 画像パスのベクトル
#' @param cluster_num クラスタ番号
#' @return フィルタされた画像パス
filter_images_by_cluster <- function(images, cluster_num) {
  if (is.null(cluster_num) || is.na(cluster_num)) {
    return(images)
  }
  
  cluster_nums <- sapply(images, extract_cluster_number)
  images[!is.na(cluster_nums) & cluster_nums == cluster_num]
}

#' Shiny用の画像ギャラリーデータを生成
#' @param images 画像パスのベクトル
#' @param max_per_page ページあたり最大画像数
#' @return ギャラリーデータ
create_gallery_data <- function(images, max_per_page = 20) {
  if (length(images) == 0) {
    return(list(
      images = character(),
      total = 0,
      pages = 0
    ))
  }
  
  list(
    images = images,
    total = length(images),
    pages = ceiling(length(images) / max_per_page),
    per_page = max_per_page
  )
}

#' 画像を更新日時順にソート（新しい順）
#' @param images 画像パスのベクトル
#' @return ソートされた画像パスのベクトル
sort_images_by_time <- function(images) {
  if (length(images) == 0) {
    return(images)
  }
  
  # 各ファイルの更新日時を取得
  file_info <- file.info(images)
  
  # 更新日時が取得できないファイルはスキップ
  valid_idx <- !is.na(file_info$mtime)
  
  if (sum(valid_idx) == 0) {
    return(images)
  }
  
  # 更新日時の新しい順にソート
  images[valid_idx][order(file_info$mtime[valid_idx], decreasing = TRUE)]
}

File: app/services/results_viewer.py
# =============================================================================
# MSI Analysis Application - Results Viewer
# 結果可視化モジュール
# =============================================================================

import re
from pathlib import Path
from typing import Optional

# 結果フォルダの主要サブディレクトリ
KEY_SUBDIRS = [
    "Harmony", "RPCA", "PCA",
    "Volcano_Plots", "Volcano_Plots_MRM",
    "Cluster_Top5_MSI", "PerCluster_Highlight",
    "RDS_Files",
]

IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg"}


def get_result_structure(result_dir: str) -> dict:
    """結果フォルダの構造を取得"""
    root = Path(result_dir)
    if not root.is_dir():
        return {"root": str(root), "subdirs": {}, "root_images": []}

    subdirs = {}
    for subdir_name in KEY_SUBDIRS:
        subdir_path = root / subdir_name
        if subdir_path.is_dir():
            images = [
                str(f) for f in subdir_path.rglob("*")
                if f.suffix.lower() in IMAGE_EXTENSIONS
            ]
            subdirs[subdir_name] = {
                "path": str(subdir_path),
                "image_count": len(images),
                "images": images,
            }

    root_images = [
        str(f) for f in root.iterdir()
        if f.is_file() and f.suffix.lower() in IMAGE_EXTENSIONS
    ]

    return {
        "root": str(root),
        "subdirs": subdirs,
        "root_images": root_images,
    }


def categorize_image(image_path: str) -> str:
    """画像カテゴリを判定"""
    filename = Path(image_path).name.lower()

    if "umap" in filename:
        return "UMAP"
    if "volcano" in filename:
        return "Volcano"
    if "msi" in filename or "top5" in filename:
        return "MSI"
    if "spatial" in filename or "cluster" in filename:
        return "Spatial"
    if "heatmap" in filename:
        return "Heatmap"
    if "tic" in filename:
        return "TIC"
    if "filter" in filename:
        return "Filtering"
    return "Other"


def organize_images_by_category(images: list[str]) -> dict[str, list[str]]:
    """画像一覧をカテゴリ別に整理"""
    if not images:
        return {}

    result: dict[str, list[str]] = {}
    for img in images:
        cat = categorize_image(img)
        result.setdefault(cat, []).append(img)
    return result


def extract_cluster_number(image_path: str) -> Optional[int]:
    """クラスタ番号を画像パスから抽出"""
    filename = Path(image_path).name

    # パターン: Cluster_0, Cluster_10 など（大文字小文字不問）
    match = re.search(r"[Cc]luster_(\d+)", filename)
    if match:
        return int(match.group(1))
    return None


def get_available_clusters(result_dir: str) -> list[int]:
    """結果フォルダ内の利用可能なクラスタ番号を取得"""
    root = Path(result_dir)
    if not root.is_dir():
        return []

    clusters = set()
    for f in root.rglob("*"):
        if f.suffix.lower() in IMAGE_EXTENSIONS:
            num = extract_cluster_number(str(f))
            if num is not None:
                clusters.add(num)
    return sorted(clusters)


def extract_sample_name(image_path: str) -> Optional[str]:
    """サンプル名を画像パスから抽出"""
    filename = Path(image_path).stem
    parts = filename.split("_")

    # 日付パターン（6桁の数字で始まる）を探す
    for i, part in enumerate(parts):
        if re.match(r"^\d{6}", part):
            return "_".join(parts[i:])
    return None


def filter_images_by_cluster(
    images: list[str], cluster_num: Optional[int]
) -> list[str]:
    """特定クラスタの画像をフィルタ"""
    if cluster_num is None:
        return images
    return [
        img for img in images
        if extract_cluster_number(img) == cluster_num
    ]


def create_gallery_data(
    images: list[str], max_per_page: int = 20
) -> dict:
    """画像ギャラリーデータを生成"""
    if not images:
        return {"images": [], "total": 0, "pages": 0, "per_page": max_per_page}

    import math
    return {
        "images": images,
        "total": len(images),
        "pages": math.ceil(len(images) / max_per_page),
        "per_page": max_per_page,
    }


def sort_images_by_time(images: list[str]) -> list[str]:
    """画像を更新日時順にソート（新しい順）"""
    if not images:
        return images

    def get_mtime(img_path: str) -> float:
        try:
            return Path(img_path).stat().st_mtime
        except OSError:
            return 0.0

    return sorted(images, key=get_mtime, reverse=True)
